Non-negative tucker with HALS

caglayantuna opened the pull request:
# HALS for Non-negative Tucker
This pull request adds a non-negative tucker with HALS algorithm to `_tucker.py` file. In our previous PR, we added `hals_nnls` function and here we use it for tucker decomposition. We added `non_negative_tucker_hals` and `initialize_tucker` function to `_tucker.py` file. Additionally, we added two nonnegative least square algorithms to proximal.py in order to update core for nonnegative tucker decomposition.

# Motivation
Using HALS for non-negative Tucker decomposition gives more accurate results according to our experiments and the literature. We compared our algorithms with the current nonnegative Tucker in Tensorly and we observed pretty good improvement. We are planning to submit an example based PR soon which will include our experiments.

# Source Modification 
* **decomposition/_tucker.py:**

First, we added `initialize_tucker` function in order to have initializing function such as `initialize_cp` for Tucker. In this function, if non_negative is True, it returns non-negative core and factors. Otherwise, it can be used for unconstrained `Tucker` decomposition as well.

We added `non_negative_tucker_hals` function which uses `hals_nnls` algorithm to calculate factors for tucker decomposition.

To compare our function, we added return_error option to `non_negative_tucker`. We also include this version in this PR.

* **tenalg/_proximal.py:**

To update core during the iteration in `non_negative_tucker_hals`, we added two fast gradient algorithms which are active set and Fast Iterative Shrinkage Thresholding Algorithm (FISTA). These functions return updated nonnegative core for `non_negative_tucker_hals`. Besides, these functions can be used for different nonnegative least square problems in the future.

* **test_tucker.py:**

Lastly, we have added a test function *"test_non_negative_tucker_hals()"* for *"non_negative_tucker_hals"* function. Simply, we have copied *"test_non_negative_tucker"* and changed related function. We have tested our function with this test function as well.

# References

1-  Bro, R., & De Jong, S. (1997). A fast non‐negativity‐constrained least squares algorithm. Journal of Chemometrics: A Journal of the Chemometrics Society, 11(5), 393-401. [Link](http://xrm.phys.northwestern.edu/research/pdf_papers/1997/bro_chemometrics_1997.pdf)

2- Beck, A., & Teboulle, M. (2009). A fast iterative shrinkage-thresholding algorithm for linear inverse problems. SIAM journal on imaging sciences, 2(1), 183-202.  [Link](https://epubs.siam.org/doi/abs/10.1137/080716542)

JeanKossaifi commented on tensorly/decomposition/_tucker.py:
```diff
@@ -294,7 +349,156 @@ def non_negative_tucker(tensor, rank, n_iter_max=10, init='svd', tol=10e-5,
+            # UtM
+            tensor_cross = multi_mode_dot(tensor, nn_factors, skip=mode, transpose=True)
+            MtU = tl.dot(unfold(tensor_cross, mode), tl.transpose(unfold(nn_core, mode)))
+            UtM = tl.transpose(MtU)
```
If we use MtU, UtM, UtU, it would be nice to have a short Notes section in the docstring to briefly introduce the notation. We can also add a short mathematical bit in the user guide.

caglayantuna replied:
I added mathematical notes to non_negative_tucker_hals function. I think these variables are mote clear now.

JeanKossaifi commented on tensorly/decomposition/_tucker.py:
```diff
@@ -294,7 +349,156 @@ def non_negative_tucker(tensor, rank, n_iter_max=10, init='svd', tol=10e-5,
+    exact : If it is True, the HALS nnls subroutines give results with high precision but it needs high computational cost. 
+        If it is False, the algorithm gives an approximate solution.
+        Default: False
+    algorithm : {'fista', 'as'}
```
I would explicitly use `'active_set'` instead of just `'as'`

caglayantuna replied:
It is 'active_set' now.

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+       Pre-computed product of the transposed of A and B.
+    pseudo_inverse: ndarray
+       Pre-computed product of the transposed of A and A.
+    x: initialized array
```
shall we call that `init`?

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+
+    Parameters
+    ----------
+    AtB: ndarray
```
Should that be Atb or ATb?

caglayantuna replied:
Now, this variable is UtM. Since M is matrix here, I kept as capital.

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+        sparse = sparsity_coefficient[-1]
+
+    if gradient_step is None:
+        gradient_step = 0.001
```
This should just be a default parameter.

caglayantuna replied:
Done

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+        norm = tl.norm(delta_x)
+        if iteration == 1:
+            norm_0 = norm
+        if norm < 0.01 * norm_0:
```
should we add a parameter `tol=0.01` ?

caglayantuna replied:
Done

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+          SIAM journal on imaging sciences, 2(1), 183-202.
+    """
+
+    if sparsity_coefficient[-1] is None:
```
why is sparsity_coefficient assumed to be a list here? 
Should we just use `sparsity_coef` to replace both sparsity_coefficient and sparse and have `sparsity_coef = 0` in the signature?

caglayantuna replied:
Now, it is in the signature as a float

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+
+    Returns
+    -------
+    x : Updated ndarray
```
```suggestion
    x : approximate solution such that Ax = b
```

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+    # Parameters
+    momentum_old = tl.tensor(1.0)
+    norm_0 = 0.0
+    x_upd = tl.copy(x)
```
We may as well spell it out, e.g. x_update

caglayantuna replied:
Ok, it is x_update now

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+        else:
+            delta_x = gradient_step * gradient
+
+        xnew = x_upd - delta_x
```
```suggestion
        x_new = x_update - gradient_step*x_gradient
```

nitpick: x_new to be consistent with rest?
I would also remove the gradient step from above and make it here for clarity.

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+        gradient = - AtB + tl.tenalg.multi_mode_dot(x_upd, pseudo_inverse, transpose=False) + sparse
+
+        if non_negative is True:
+            delta_x = tl.where(gradient_step * gradient < x, gradient_step * gradient, x_upd)
```
Why not update gradient instead of introducing delta_x? We no longer need gradient after this. 
Generally it's good to introduce as few variables as needed to limit memory use and improve readability.

caglayantuna replied:
The idea was to handle non_negativity easily. But as you said, it is possible to do that without it. There is no delta_x now.

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+    x_upd = tl.copy(x)
+
+    for iteration in range(n_iter_max):
+        gradient = - AtB + tl.tenalg.multi_mode_dot(x_upd, pseudo_inverse, transpose=False) + sparse
```
x_gradient?

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+        Default: 100
+    non_negative : bool, default is False
+                   if True, result will be non-negative
+    gradient_step : float
```
we could call this learning_rate or lr for consistency with opt. frameworks?

caglayantuna replied:
Now it is "lr"

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+    second = tl.dot(AtA, x_vec)
+    w = first - second
+    P = x_vec > 0
+    R = x_vec <= 0
```
Could we find an explicit name for P and R (e.g. positive_set ...)? I would remove first and second and directly use their value, and rename w to something more explicit as well.

caglayantuna replied:
Now, their names are changed to passive_set and active_set.

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+
+    for iteration in range(n_iter_max):
+        # Early stopping criterion
+        if True in R and tl.max(w[R]) <= tol:
```
I think we can add an `any` function to the backend and use `if tl.any(R)`?

caglayantuna replied:
yes, it could be better

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+
+        # gradient update
+        second = tl.dot(AtA, x_vec)
+        w = first - second
```
same as above, no need to introduce second as it's not used anywhere else, just 
`AtB - tl.dot(AtA, x_vec)`

caglayantuna replied:
There is no second or first anymore

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+                        else:
+                            x_vec = tl.index_update(x_vec, tl.index[int(i)], 0)
+                    x_vec = tl.clip(x_vec, 0, tl.max(x_vec))
+                except:
```
What errors are we catching? We should add a specific exception here and not catch everything (e.g. user may be sending `KeyboardInterrupt`.

caglayantuna replied:
It is numpy.linalg.LinAlgError. I dont know which exception covers that error.

cohenjer replied:
It's basically when the linear system is ill-conditionned or has infinitely many solutions. Not sure how to catch this either, some idea?

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+            indice = tl.argmax(w)
+            P = tl.index_update(P, tl.index[indice], True)
+            R = tl.index_update(R, tl.index[indice], False)
+        # To avoid singularity error when initial x exists
```
Is that a single occurrence, can we test for that?

caglayantuna replied:
When initial x exists, we may have this error. According to the my tests, it occurs when all the P (changed, now passive set) is True. So, it happens once because when P has only true, R will have only false then iteration stops in the end.

JeanKossaifi replied:
So we could test with tl.any?

JeanKossaifi reviewed (changes requested):
Thanks @caglayantuna, another great PR!

It would be great to have a pass focusing on making the code easy to read with more explicit (and where possible, less) variables, particularly for the active set solver. 
Would be good to have a small unit-test for the solvers as well.

caglayantuna commented on tensorly/decomposition/_tucker.py:
```diff
@@ -1,18 +1,73 @@
+    else:
+        (core, factors) = init
+ 
+    if non_negative is True:
```
Now, make_svd_non_negative is used to initialize. Should we add this function to init?

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+
+def active_set_nnls(AtA, AtB, x=None, n_iter_max=100, tol=10e-8):
+    """
+     Active set algorithm for non-negative least square solution.
```
Can we also describe briefly the problem solved - otherwise the AtA and AtB, combined with x being None by default may be opaque for users (though admittedly they are not that likely to use this function directly).

caglayantuna commented on tensorly/decomposition/_tucker.py:
```diff
@@ -294,7 +349,156 @@ def non_negative_tucker(tensor, rank, n_iter_max=10, init='svd', tol=10e-5,
+        # updating core
+        if algorithm == 'fista':
+            pseudo_inverse[-1] = tl.dot(tl.transpose(nn_factors[-1]), nn_factors[-1])  # all_MtM
+            AtB = multi_mode_dot(tensor, nn_factors, transpose=True)
```
hey, names are clear in the current version. I am using only UTM and UtU now. For instance, AtB is core_estimation

caglayantuna commented:
Hey @JeanKossaifi, thanks for your review. I updated the PR according to your comments. You could find my answers to your comments. 

I have added unit tests for each solver however I have a problem about tensor_to_vec function in test_proximal file. I couldn't understand why it returns error in there although I use the same function elsewhere without any error. Also, I think these solver tests are not perfect because they are not passing the test always. If you have any suggestion, it would be perfect.

JeanKossaifi commented:
Thanks @caglayantuna, this looks good!
Re. the test failure, you should use directly `tl.tensor_to_vec` (`base` is not automatically imported, if you want to use tensorly.base you have to import it first). 

I've also added `any` to the backend so we can use it now!

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,196 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+            active_set = tl.index_update(active_set, tl.index[indice], False)
+        # To avoid singularity error when initial x exists
+        try:
+            a = tl.tensor(tl.to_numpy(UtU)[passive_set, :][:, passive_set], **tl.context(UtU))
```
I've been debating about this and trying to make it work - the main issue is TensorFlow: I am thinking to either not support tensorflow here or checking for the backend and having a separate case for it. What do you think @caglayantuna @cohenjer 

More generally I'm more and more inclined to phase out tensorflow support (we have jax anyway).

caglayantuna replied:
Yes I agree, it is difficult to handle tensorflow for some situations. I changed the codes as you suggested. As a suggestion, seperate case could be directing active set to fista with an informative warning if backend is tensorflow.

cohenjer replied:
It does seem like the Tensorflow backend is causing us a lot of problems. Not sure if people are actually using it? Maybe a quick survey on this backend usage would help checking if it is worth the effort to support it at the moment?

JeanKossaifi replied:
Yes, I'll create a survey. @caglayantuna that sounds like a good solution at least for now - we can just make it work for backends other than tensorflow and at the beginning of the nn_tucker, if method is 'fista' then check for the backend and return an error if is tensorflow.

JeanKossaifi commented on tensorly/tenalg/proximal.py:
```diff
@@ -248,3 +248,196 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+
+                if True not in passive_set or tl.min(s[passive_set]) > 0:
+                    break
+        # set x to s
```
Could we also find a more explicit name for `s` (and `c` and `a`?) 
The other names we now use make the code much easier to read (passive/active_set, x_gradient, etc)!

caglayantuna replied:
I changed c to passive_solution, s to support_vec and removed a by merging that calculation with the next line.

caglayantuna commented:
Hey @JeanKossaifi , thanks for "any" function, it is better for Tensorly. I realized the problem with tensor_to_vec function. In test_proximal file, tensorly is imported from backend as T. I think either we should add tensor_to_vec to backend init file or we should import tensorly as usual. 

Additionally, I added exact definition to docstring of hals because it was missing. Besides, I have changed the parameters of hals when exact is true in order to pass test.

JeanKossaifi commented:
Looks great overall - just wondering whether we can find a way to remove the try/except. 
We also need to add this to the doc. Would be great to have a small example to e.g. compare regular decomposition vs NN multiplicative updates vs NN via HALS!

caglayantuna commented:
Thanks for your comment. We were discussing with Jeremy today and we think that there is no any other solution than try except for active set solution. Because, there could be a singularity error depending on the initial x especially when higher ranks are selected. So, I can't define a specific reason to avoid that. Related to example, I have almost finished preparing notebooks for both cp and tucker nn. [Link for notebooks](https://github.com/caglayantuna/tensorly-notebooks/tree/non-negative/02_tensor_decomposition). I will submit these examples after this tucker HALS PR merged.

JeanKossaifi commented:
Awesome, thanks @caglayantuna and @cohenjer, merging!
